rclcpp/include/rclcpp/node_impl.hpp

```diff
@@ -206,18 +145,18 @@ Node::create_subscription(
   typename rclcpp::message_memory_strategy::MessageMemoryStrategy<
     typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>::SharedPtr
   msg_mem_strat,
-  std::shared_ptr<Alloc> allocator,
-  IntraProcessSetting use_intra_process_comm)
+  std::shared_ptr<Alloc> allocator)
 {
+  rclcpp::QoS qos(rclcpp::QoSInitialization::from_rmw(qos_profile));
+  qos.get_rmw_qos_profile() = qos_profile;
```

Comment:
Here we're overriding the special logic in the QoS constructor. In case the history is set to `RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT` or `RMW_QOS_POLICY_HISTORY_UNKNOWN`, it won't default to `KeepAll`. I think this should be replaced with:

```c++
  rclcpp::QoS qos(rclcpp::QoSInitialization::from_rmw(qos_profile), qos_profile);
```

This occurs in a few other places following the use of `QoSInitialization::from_rmw`.

Reply:
The reason this doesn't work is that `from_rmw` doesn't handle `RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT` or `RMW_QOS_POLICY_HISTORY_UNKNOWN` properly, and so by doing `qos.get_rmw_qos_profile() = qos_profile` after construction we ensure that we store all the settings from `qos_profile`, and the initialization I use is just to successfully create the QoS in the first place, it doesn't actually matter. In fact in at least one place I think I initialized with `rclcpp::KeepAll()` just to make it shorter, before I decided to use the `from_rmw` in a few places.

This is super awkward, I know, but it's also mostly only to support deprecated use cases. So I expect cases like this to go away eventually.